trust-store-trust-policy-specification.md

```diff
@@ -97,27 +97,46 @@ The trust policy is represented as JSON data structure as shown below:
 ```json
 {
     "version": "1.0",
-    "trustPolicies": {
-        "name": "trust-policy-name",
-        "trustStores": [ "trust-store-name-1", "trust-store-name-2" ],
-        "expiryValidations": {
-            "signatureExpiry": "enforce/warn",
-            "signingIdentityExpiry": "enforce/warn",
-            "timestampExpiry": "enforce/warn"
+    "trustPolicies": [
+        {
+            "name": "trust-policy-name-1",
+            "Scopes": ["wabbit-networks.io/software"],
+            "trustStores": [ "trust-store-name-1", "trust-store-name-2" ],
+            "expiryValidations": {
+                "signatureExpiry": "enforce/warn",
+                "signingIdentityExpiry": "enforce/warn",
+                "timestampExpiry": "enforce/warn"
+            },
+            "revocationValidations": {
+                "signingIdentityRevocation": "enforceWithFailOpen/enforceWithFailClose/warn/skip",
+                "timestampRevocation": "enforceWithFailOpen/enforceWithFailClose/warn/skip"
+            }
         },
-        "revocationValidations": {
-            "signingIdentityRevocation": "enforceWithFailOpen/enforceWithFailClose/warn/skip",
-            "timestampRevocation": "enforceWithFailOpen/enforceWithFailClose/warn/skip"
+        {
+            "name": "trust-policy-name-2",
+            "trustStores": [ "trust-store-name-1", "trust-store-name-2" ],
+            "expiryValidations": {
+                "signatureExpiry": "enforce/warn",
+                "signingIdentityExpiry": "enforce/warn",
+                "timestampExpiry": "enforce/warn"
+            },
+            "revocationValidations": {
+                "signingIdentityRevocation": "enforceWithFailOpen/enforceWithFailClose/warn/skip",
+                "timestampRevocation": "enforceWithFailOpen/enforceWithFailClose/warn/skip"
+            }
         }
-    }
+
+    ]
 }
 ```
 
 Property descriptions
 
 - **`version`**(*string*): This REQUIRED property is the version of the trust policy. The supported value is `1.0`.
-- **`trustPolicies`**(*string-object map*): This REQUIRED property represents a collection of trust policies. As of now Notary v2 only supports one trust policy.
-  - **`name`**(*string*): The name of the trust policy
+- **`trustPolicies`**(*string-array of objects map*): This REQUIRED property represents a collection of trust policies.
+  - **`name`**(*string*): The name of the trust policy.
+  - **`scope`**(*array of string*): The scope determines which trust policy is applicable for a given artifact. The scope field supports array of prefix-based filtering on registry-name/namespace+repository-name. For an artifact, if there is no applicable trust policy, then signature evaluation must be skipped.
```

Comment:
By my reading, the examples seem to assume that registry name and namespace+repo-name are all whole path components. Is that enforced by the prefix matching logic?

For example, given `"Scopes": ["wabbit-networks.io/software"]`, is `wabbit-networks.io/softwareasdfbadguy` considered to be applicable, because it matches the given prefix? Or is there an implicit trailing `/` in that scope that makes `software` a namespace and only `...software/asdf` should match? If the scope is `wabbit-networks.io`, is `wabbit-networks.io.badguyinc.com` matching?

If the `/` is implicitly assumed, I might consider making it a requirement of the scope to include an explicit trailing `/`, at least under circumstances where it may be ambiguous, which may be all of them.

Reply:
Also, if scope does normal reference expansion, is `registry:5000` a hostname/port, or an image name/tag on Docker Hub? It gets complicated distinguishing the two when there's a partial string. We may need a hard requirement on no short references in scopes, everything must be explicit `docker.io/library/registry` rather than `registry`.

Reply:
@imjasonh 

`/` is not implicitly assumed, its plain longest prefix match.
The scope will be used by a user to define trusted signing-identities and verification behavior for the registry(s), namespace(s), and repository(s). So even if `Scopes": ["wabbit-networks.io/software"]` matches `wabbit-networks.io/softwareasdfbadguy`, it just means that artifacts from `softwareasdfbadguy` will be trusted only if they are signed using trusted identities mentioned in policy associated with `wabbit-networks.io/software` scope.

Am I missing something?

---
@sudo-bmitch 
No, their wont be any reference expansion on the scope, repository URIs must be explicit. Will add this to scope constraits.

Reply:
I have a concern of rollback attack with expired signatures due to prefix match. Is that possible?

Reply:
I think we're saying the namespace must be complete, to avoid repo squatting as pointed out by @imjasonh 
I believe we're saying:

"Scopes": ["wabbit-networks.io/software"], would NOT apply to `wabbit-networks.io/softwareasdfbadguy`

`wabbit-networks.io/softwareasdfbadguy` is a different namespace than `/software`

I don't think we need to require a trailing `/` as that would be an easy security hole.

In fact, I'd suggest we shouldn't use * as valid characters. Users specify a root namespace. Anything under is applied to the scope, using the longest namespace pattern outlined.

Reply:
So I have updated the evaluation approach:
- If there is no trailing `/` then we consider it as repository scope i.e. this scope will be applicable only if it matches repository URI
- If there is trailing `/` then we consider scope as namespace scope i.e. this scope will be applicable only if it matches complete or partial namespace of registry.

For example, in the case of the `wabbit-networks.io/software/ocp-release` repository, the possible matching scopes are
-  `wabbit-networks.io/software/ocp-release`
- `wabbit-networks.io/software/`
- `wabbit-networks.io/`
and  the priority order will be `wabbit-networks.io/software/ocp-release` > `wabbit-networks.io/software/` > `wabbit-networks.io/`.

Reply:
This solves the repo squatting, awesome, thanks.
For the `/` or not, this sounds like a bug farm, as users will sometimes forget to add the `/`, or they're using variables or substring parsing logic that may or may not have it. 

Can we simply treat the a named element as complete or not? In other words:
`wabbit-networks.io/software/` and `wabbit-networks.io/software` are equivalent.

Reply:
The problem with considering `wabbit-networks.io/software/` and `wabbit-networks.io/software` is that it doesn't allows user to specify different policy for namespace `wabbit-networks.io/software/` and repository `wabbit-networks.io/software`.